JavaScript/Angular.md

```diff
@@ -0,0 +1,1551 @@
+
+To convert an observable to a promise, just call `obs$.toPromise()` and then you can await it.
+
+One important thing to note is that you are awaiting for **completion** and not for **emission** of values. Knowing that, what do you expect the result to be in the following example?
+
+```ts
+const source$: Subject<string> = new Subject();
+const sourcePromise: Promise<string> = source$.toPromise();
+
+source$.next('foo');
+source$.next('bar');
+source$.complete();
+
+const result = await sourcePromise;
+```
+
+Spoiler alert, the result will be `bar`, and if we do not complete the source observable, awaiting will hang indefinitely.
+
+## File/Module Organization and Naming
+
+If you created the app using Angular CLI, there is already some clear structure in place. Building on that structure, you will probably want to add some more dirs to help you organize your code. Here are some guidelines:
+
+- Try to split up your application into modules which you can reuse. If, for example, you have to handle lazy loading of all images in the app, create a `Image` module and component in `app/components` and use it in other modules/components whenever you have to show an image. Export the declared component and import/provide dependencies which your component uses.
+
+- Container components which are loaded directly via routing (by either `component` or `loadChildren`) should be placed in `app/pages` dir. If a component is loaded because you have used it in a template, that component should be in `app/components`.
+  - One little exception here: if there is some very specific component tied to some routing submodule which is used only in that submodule, it is OK to leave it in that submodule (but as soon as it is used in another routing submodule, move it to `app/components`).
+
+- Store all your animations in `app/animations/` and import them for usage in component decorators. If you want to have some animation properties like duration configurable, you can export "animation factory" function which returns the actual `AnimationTriggerMetadata`.
+
+- Classes which might be used throughout the application (e.g. generic `Dictionary` class) should be in `app/classes`.
+
+- Entity models (`User`, `Post`, etc.) should be in `app/models`.
+
+- If you are using some JS libs which do not have typings, add your own custom typing in `app/custom-typings`.
+
+- Whenever you have some attribute whose value can be some value from a set of values, make an enum and store that enum in `app/enums`. Good example is an enum for HTTP status codes. If you need some metadata for the enum, create an enum data file which exports what is basically a dictionary with enum values used as keys and use whatever you need for values. One such example would be: `export const httpStatusCodeData = { [HttpStatus.OK_200]: { translationKey: 'http.success' } }`
```

Comment:
I found that it's beneficial to define an interface for enum data object element, as it helps when resolving that data in the code. For example:

```
export enum HttpStatus {
  OK_200: 200,
  NOTFOUND_404: 404,
}

export interface IHttpStatusCodeData {
  translationKey: string;
}

export const httpStatusCodeData: Record<HttpStatus, IHttpStatusCodeData> = { 
  [HttpStatus.OK_200]: { 
    translationKey: 'http.success',
  }, 
  [HttpStatus.NOTFOUND_404]: { 
    translationKey: 'http.notfound',
  },
}
```

Reply:
I Mentioned that adding an interface can be a good idea, but I didn't add this whole example because it didn't really fit into this list of bullet points where no other bullet had a multi-line code snippet